apt/method.go

```diff
@@ -0,0 +1,225 @@
+	"os"
+	"strings"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+)
+
+func NewAptMethod(output io.Writer, input *bufio.Reader) *AptMethod {
+	return &AptMethod{
+		config: &AptMethodConfig{},
+		writer: NewAptMessageWriter(output),
+		reader: NewAptMessageReader(input),
+	}
+}
+
+// httpClient exists to enable mocking of http.Client.
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// downloader exists to enable mocking of AptMethod.download.
+type downloader interface {
+	download(io.ReadCloser, string) (string, error)
+}
+
+type realDownloader struct{}
+
+type AptMethod struct {
+	config *AptMethodConfig
+	writer *AptMessageWriter
+	reader *AptMessageReader
+	client httpClient
+	dl     downloader
+}
+
+type AptMethodConfig struct {
+	serviceAccountJSON, serviceAccountEmail string
+}
+
+func (m *AptMethod) Run(ctx context.Context) {
+	m.writer.SendCapabilities()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+		msg, err := m.reader.ReadMessage(ctx)
+		if err != nil {
+			// TODO: log?
+			continue
+		}
+		switch msg.code {
+		case 600:
+			m.handleAcquire(msg)
+		case 601:
+			m.handleConfigure(msg)
+		default:
+			m.writer.Fail("Unsupported API")
+		}
+	}
+}
+
+func (m *AptMethod) getClient() error {
+	if m.client != nil {
+		return nil
+	}
+
+	var ts oauth2.TokenSource
+	ctx := context.Background()
+	switch {
+	case m.config.serviceAccountJSON != "":
+		json, err := ioutil.ReadFile(m.config.serviceAccountJSON)
+		if err == nil {
+			creds, ierr := google.CredentialsFromJSON(ctx, json)
+			if ierr == nil {
+				ts = creds.TokenSource
+			}
+		}
+	case m.config.serviceAccountEmail != "":
+		ts = google.ComputeTokenSource(m.config.serviceAccountEmail)
+	default:
+		creds, err := google.FindDefaultCredentials(ctx)
+		if err == nil {
+			ts = creds.TokenSource
+		}
+	}
+	if ts == nil {
+		return errors.New("Failed to obtain token source")
+	}
+	m.client = oauth2.NewClient(ctx, ts)
+	return nil
+}
+
+func (r realDownloader) download(body io.ReadCloser, filename string) (string, error) {
+	defer body.Close()
+	data, err := ioutil.ReadAll(body)
+	if err != nil {
+		return "", err
+	}
+	file, err := os.Create(filename)
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+
+	_, err = file.Write(data)
+	return fmt.Sprintf("%x", md5.Sum(data)), err
+}
+
+// TODO: testing. this validates the message (could be separated), optionally adds a header, makes a GET request, and sends appropriate 2xx messages to apt.
+// it also returns an error that we don't care about..
+func (m *AptMethod) handleAcquire(msg *AptMessage) error {
+	// === VALIDATE MESSAGE ===
+	uri := msg.Get("URI")
+	if uri == "" {
+		err := errors.New("No URI provided")
+		m.writer.Fail(err.Error())
+		return err
+	}
+	filename := msg.Get("Filename")
+	if filename == "" {
+		err := errors.New("No filename provided")
+		m.writer.FailURI(uri, err.Error())
+		return err
+	}
+	ifModifiedSince := msg.Get("Last-Modified")
```

Comment:
this doesn't really belog to validate as it's not used for anything till makeRequest